docs/IWYUStyle.md

```diff
@@ -0,0 +1,242 @@
+by the language construct or IWYU feature they cover. This tends to produce a
+large number of testcases, and we'll probably have to go back and organize them
+further at some point.
+
+Some things to keep in mind:
+
+* Keep testcase filenames short -- this is a case of "as short as possible, but
+  no shorter". Every file has its name in the license header, the stem is
+  repeated for supporting headers, and testcase names occasionally need to be
+  typed without completion, so it's nice if they're short and sweet. But they
+  also need to be meaningful. Find the balance.
+* If a change pertains to a feature covered by an existing testcase, consider
+  extending it instead of creating a new one.
+* Don't use another testcase's supporting headers; that creates unwanted
+  coupling between tests. But see below for a set of headers we do encourage
+  sharing.
+
+### Reusable headers ###
+
+Prefer the reusable headers for simple constructs.
+
+The test stack contains a few headers aimed for reuse:
+
+* `indirect.h` -- defines `IndirectClass`, a very basic C++ class
+* `direct.h` -- includes `indirect.h`
+* `subdir/indirect_subdir.h` -- defines `IndirectSubDirClass`
+* `subdir/direct_subdir.h` -- includes `subdir/indirect_subdir.h`
+* `subdir/dotdot_indirect.h` -- includes `../indirect.h`
+
+If you need a basic class as a return type, argument type, variable or template
+argument, `IndirectClass` should be your preference.
+
+The `direct*.h` variants are used to force IWYU to do its analysis and
+replacement; it will suggest that you remove `direct.h` and include `indirect.h`
+directly. More on this below.
+
+The `subdir/` headers can be used to test that IWYU is not confused by directory
+structure (this has been known to happen).
+
+### Avoid system headers ###
+
+Avoid using system headers for regression tests.
+
+It's tempting to phrase your testcase in terms of the library that triggered a
+bug. But our test runners do not have all libraries installed. The C++ standard
+library is a non-obvious example of this: it's always available, but IWYU aims
+to be portable, and standard library _implementations_ differ wildly.
+
+For example, if `std::map` triggers a bad behavior with IWYU on your MSVC system
+because it uses some template trick in its implementation, it's likely libstdc++
+or libc++ do not use the same trick, and exhibit different behavior.
+
+Likewise, the internal physical design details of standard library
+implementations are usually wildly different -- libstdc++ has all its private
+headers in a directory called `bits/`, MSVC prefixes all private library headers
+with `x`, e.g. `xstring`, etc. IWYU uses mappings to overcome these differences.
+
+Since we can't control what standard library IWYU will be tested with, we prefer
+to avoid standard library types and headers as far as possible, and instead try
+to reduce constructs that cause problems to a minimal mockup. Clang takes a
+similar approach in its test suite.
```

Comment:
It might be worth mentioning that `badinc.cc` does use the standard library and due to this, it is currently broken on MSVC.

Also, another reason to avoid system headers is performance. Running IWYU on debug mode and waiting for a minute to process thoundands of lines of library code is no fun :)